### ISSUE TITLE
Rolling update config checker

### ISSUE THREAD

**Issue by patriknw** [akka/akka#24009]
It would be very useful to be able to check if config is compatible when joining a cluster node in rolling upgrade scenarios.

A specific example is that the Cluster Sharding ddata mode is not compatible with the persistence mode.

We probably need to amend the Join and Welcome message to be able to pass config. Good if check can be done on both sides. If Welcome message is too late, then `InitJoinAck` might be better.

**Comment by renatocaval** [comment]
I have done some preliminary investigations. Here is my understanding so far. 

If we extend the `Join` message to pass some configuration we will have a situation where a new node sends a `Join` (with config) message to an existing cluster that is running a previous version.

Because the existing cluster is on a previous version, it won't be able to deserialize the config part of the `Join` message. And it won't be able to full-fill the protocol my acknowledging that it's OK or NOK to join that cluster.  As a result, it will simply send an `InitJoinAck` like it would normally do. 

When receiving the `InitJoinAck`, the new node won't get any config back from old nodes, so it can't decide if it's safe to continue or not. The only way to solve it is by having a configuration instructing the new node how to proceed. If it doesn't receive any config information from the seed-node, it simply gives up and shutdown, or it ignores the fact and joins anyway. This could be done with a config like:  `enforce-config-check-when-joining = false`

By default, new nodes must simply take the risk and join an existing cluster without knowing if their configuration (shard store or others) are compatible. The other option is to abort the operation and not join, but that means that we will force a cluster shutdown. 

If we add this feature, first we will have to update all nodes to support the new protocol and then be able to use it. Once we have them all using the same version, we could set `enforce-config-check-when-joining = true`. 

I'm wondering how much we gain from that. From a Lagom perspective, that may help because we can set `enforce-config-check-when-joining = true` and force a cluster shutdown or at least make the user aware of this and force them to tweak the config. But again we rely on people reading the documentation and/or migration guide. While the whole idea was to have something that would protect unadvised users. 

Anyway, I would like to know if someone has any other thoughts or strategy to tackle this. My current feeling is that the added value is low because it will only work properly if the end users follow the documentation and migration guide.

**Comment by patriknw** [comment]
We should probably backport the feature to Akka 2.4 to allow safe migrations between 2.4 and 2.5, i.e. they must first upgrade to Akka 2.4.21 and then to 2.5.x.

If a Lagom 1.4 node tries to join an old 1.3 node (without this feature) the 1.4 node will not receive the config and will then not join. In that we can enforce that they first upgrade to the new 1.3 version (with this feature).

**Comment by renatocaval** [comment]
Ok, if I understand you correctly the goal is them to force the failure so we can later enforce the protection. 

In that case, we must still provide the configuration flag to be able to bypass it. 

As such, the message that we sent is: you can't simply upgrade to those versions unless you explicitly set the flag to false (`enforce-config-check-when-joining = false`). In which case, you must be sure that the sharding store is either both `ddata` or both `persistence`. 

Basically, we block them before they corrupt their data and we ask to be explicit. 

Is that what you had in mind?

**Comment by TimMoore** [comment]
If we keep the default configuration for Lagom 1.4.x to use persistence as the shard coordination mode, then IMHO it's less of a concern whether Akka 2.4 supports this or not.

I think this is more of an issue when we do change the default value in a upcoming version (1.5 or 2.0). At that point, we can require that people migrate from the latest Lagom 1.4 version, which will include the config exchange.

If we fast-track the change and get it into Lagom 1.3.x and 1.4.0, then we could potentially change the default in 1.4.0, however we are also anticipating another migration requiring downtime for the upcoming Akka Persistence Cassandra schema change, so I'd rather batch them into one disruptive release, rather than make every release disruptive. Furthermore, 1.4.0 is very close to complete, and I would prefer not to add another blocking dependency.

**Comment by patriknw** [comment]
I agree with @TimMoore .

I think this is a great feature for the future (Akka 2.5 -> 2.6) but it shouldn't block the Lagom release.
To make this right it requires some work, e.g. it should also be possibly for external modules to register validation rules, e.g. Lagom can have it's own rules verifying Lagom specific configuration.
